// testdouble/lib/util/lodash-wrap.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('lodash/assign');

var _assign2 = _interopRequireDefault(_assign);

var _capitalize = require('lodash/capitalize');

var _capitalize2 = _interopRequireDefault(_capitalize);

var _clone = require('lodash/clone');

var _clone2 = _interopRequireDefault(_clone);

var _compact = require('lodash/compact');

var _compact2 = _interopRequireDefault(_compact);

var _defer = require('lodash/defer');

var _defer2 = _interopRequireDefault(_defer);

var _delay = require('lodash/delay');

var _delay2 = _interopRequireDefault(_delay);

var _each = require('lodash/each');

var _each2 = _interopRequireDefault(_each);

var _every = require('lodash/every');

var _every2 = _interopRequireDefault(_every);

var _extend = require('lodash/extend');

var _extend2 = _interopRequireDefault(_extend);

var _filter = require('lodash/filter');

var _filter2 = _interopRequireDefault(_filter);

var _find = require('lodash/find');

var _find2 = _interopRequireDefault(_find);

var _findLast = require('lodash/findLast');

var _findLast2 = _interopRequireDefault(_findLast);

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _groupBy = require('lodash/groupBy');

var _groupBy2 = _interopRequireDefault(_groupBy);

var _includes = require('lodash/includes');

var _includes2 = _interopRequireDefault(_includes);

var _invoke = require('lodash/invoke');

var _invoke2 = _interopRequireDefault(_invoke);

var _isArguments = require('lodash/isArguments');

var _isArguments2 = _interopRequireDefault(_isArguments);

var _isArray = require('lodash/isArray');

var _isArray2 = _interopRequireDefault(_isArray);

var _isBoolean = require('lodash/isBoolean');

var _isBoolean2 = _interopRequireDefault(_isBoolean);

var _isEmpty = require('lodash/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _isEqual = require('lodash/isEqual');

var _isEqual2 = _interopRequireDefault(_isEqual);

var _isEqualWith = require('lodash/isEqualWith');

var _isEqualWith2 = _interopRequireDefault(_isEqualWith);

var _isFunction = require('lodash/isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

var _isNumber = require('lodash/isNumber');

var _isNumber2 = _interopRequireDefault(_isNumber);

var _isObject = require('lodash/isObject');

var _isObject2 = _interopRequireDefault(_isObject);

var _isObjectLike = require('lodash/isObjectLike');

var _isObjectLike2 = _interopRequireDefault(_isObjectLike);

var _isRegExp = require('lodash/isRegExp');

var _isRegExp2 = _interopRequireDefault(_isRegExp);

var _isString = require('lodash/isString');

var _isString2 = _interopRequireDefault(_isString);

var _keys = require('lodash/keys');

var _keys2 = _interopRequireDefault(_keys);

var _last = require('lodash/last');

var _last2 = _interopRequireDefault(_last);

var _map = require('lodash/map');

var _map2 = _interopRequireDefault(_map);

var _reduce = require('lodash/reduce');

var _reduce2 = _interopRequireDefault(_reduce);

var _reject = require('lodash/reject');

var _reject2 = _interopRequireDefault(_reject);

var _some = require('lodash/some');

var _some2 = _interopRequireDefault(_some);

var _tap = require('lodash/tap');

var _tap2 = _interopRequireDefault(_tap);

var _toArray = require('lodash/toArray');

var _toArray2 = _interopRequireDefault(_toArray);

var _transform = require('lodash/transform');

var _transform2 = _interopRequireDefault(_transform);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  assign: _assign2.default,
  capitalize: _capitalize2.default,
  clone: _clone2.default,
  compact: _compact2.default,
  defer: _defer2.default,
  delay: _delay2.default,
  each: _each2.default,
  every: _every2.default,
  extend: _extend2.default,
  filter: _filter2.default,
  find: _find2.default,
  findLast: _findLast2.default,
  get: _get2.default,
  groupBy: _groupBy2.default,
  includes: _includes2.default,
  invoke: _invoke2.default,
  isArguments: _isArguments2.default,
  isArray: _isArray2.default,
  isBoolean: _isBoolean2.default,
  isEmpty: _isEmpty2.default,
  isEqual: _isEqual2.default,
  isEqualWith: _isEqualWith2.default,
  isFunction: _isFunction2.default,
  isNumber: _isNumber2.default,
  isObject: _isObject2.default,
  isObjectLike: _isObjectLike2.default,
  isRegExp: _isRegExp2.default,
  isString: _isString2.default,
  keys: _keys2.default,
  last: _last2.default,
  map: _map2.default,
  reduce: _reduce2.default,
  reject: _reject2.default,
  some: _some2.default,
  tap: _tap2.default,
  toArray: _toArray2.default,
  transform: _transform2.default
};
