// testdouble/lib/matchers/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

var _captor = require('./builtin/captor');

var _captor2 = _interopRequireDefault(_captor);

var _isA = require('./builtin/is-a');

var _isA2 = _interopRequireDefault(_isA);

var _contains = require('./builtin/contains');

var _contains2 = _interopRequireDefault(_contains);

var _anything = require('./builtin/anything');

var _anything2 = _interopRequireDefault(_anything);

var _argThat = require('./builtin/arg-that');

var _argThat2 = _interopRequireDefault(_argThat);

var _not = require('./builtin/not');

var _not2 = _interopRequireDefault(_not);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  create: _create2.default,
  captor: _captor2.default,
  isA: _isA2.default,
  anything: _anything2.default,
  contains: _contains2.default,
  argThat: _argThat2.default,
  not: _not2.default
};
