// testdouble/lib/index.js
'use strict';

var _function = require('./function');

var _function2 = _interopRequireDefault(_function);

var _object = require('./object');

var _object2 = _interopRequireDefault(_object);

var _constructor = require('./constructor');

var _constructor2 = _interopRequireDefault(_constructor);

var _when = require('./when');

var _when2 = _interopRequireDefault(_when);

var _verify = require('./verify');

var _verify2 = _interopRequireDefault(_verify);

var _matchers = require('./matchers');

var _matchers2 = _interopRequireDefault(_matchers);

var _replace = require('./replace');

var _replace2 = _interopRequireDefault(_replace);

var _explain = require('./explain');

var _explain2 = _interopRequireDefault(_explain);

var _reset = require('./reset');

var _reset2 = _interopRequireDefault(_reset);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _callback = require('./callback');

var _callback2 = _interopRequireDefault(_callback);

var _version = require('./version');

var _version2 = _interopRequireDefault(_version);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = {
  function: _function2.default,
  func: _function2.default,
  object: _object2.default,
  constructor: _constructor2.default,
  when: _when2.default,
  verify: _verify2.default,
  matchers: _matchers2.default,
  replace: _replace2.default,
  explain: _explain2.default,
  reset: _reset2.default,
  config: _config2.default,
  callback: _callback2.default,
  version: _version2.default
};
